onnxruntime/core/graph/graph.cc

```diff
@@ -2377,12 +2368,36 @@ Status Graph::SetGraphInputsOutputs() {
       }
     }
 
-    // Set graph outputs
-    auto end = graph_output_args.end();
-    for (auto& name : ordered_output_names) {
-      auto graph_output = graph_output_args.find(name);
-      if (graph_output != end) {
-        graph_outputs_.push_back(graph_output->second);
+    if (!graph_outputs_manually_set) {
+      // Set graph outputs in order.
+      std::vector<size_t> graph_output_args_index;
+      for (auto output_arg : graph_output_args) {
+        graph_output_args_index.push_back(output_arg.second);
+      }
+      std::sort(graph_output_args_index.begin(), graph_output_args_index.end());
+      for (auto& output_arg_index : graph_output_args_index) {
+        graph_outputs_.push_back(output_node_args_in_order[output_arg_index]);
+      }
+    } else {
+      // Validation: a graph output must be specified in graph inputs, initializers or nodes' outputs.
+      auto& outputs = GetOutputs();
```

Comment:
Not sure this is the correct validation. Assuming the NodeArg in the manual list comes from this graph, doesn't that equate to it being in the inputs/initializers/node outputs?

In terms of validation I was thinking more along the lines that all the 'real' outputs had to be in the manual list. Or do you want to allow removing a real output via the manual list?

Reply:
why this "all the 'real' outputs had to be in the manual list"?

"doesn't that equate to it being in the inputs/initializers/node outputs?" this may not be true sometimes after optimization, I think. of course, it's ok for me to remove this kind of validation.